common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+// Lacking that testing infrastructure and a full Windows port, the operations
+// here are manually compared with LLVM's filesystem library to ensure a
+// reasonable Windows implementation is possible.
+//
+// The library uses C++'s `std::filesystem::path` as its abstraction for paths,
+// and provides two core APIs: open directories and files.
+//
+// Open directories provide relative- and absolute-path based operations to open
+// other directories or files. This allows secure creation of directories even
+// in the face of adversarial operations for example in a shared `/tmp`
+// directory. There is a `constexpr` current working directory available as
+// `Cwd()` that models normal filesystem operations with paths.
+//
+// Open files provide read, write, and other operations on the file. There are
+// separate types for read-only, write-only, and read-write files to model the
+// different APIs available.
+//
+// The APIs for both directories and files are primarily on `*Ref` types that
+// model a non-owning reference to the directory or file. These types are the
+// preferred types to use on an API boundary. Owning versions are provided that
+// ensure the file or directory is closed on destruction. Files support explicit
+// closing in order to observe any close-specific errors.
+//
+// Where APIs require flag parameters of some form, this library provides
+// enumerations that model those flags. The enumeration values are in turn
+// chosen to simplify passing these to specific native APIs. This means the
+// enumeration *values* should not be expected to be portable across platforms.
+// Customizing the values is part of the larger TODO to port the implementation
+// to Windows.
+namespace Carbon::Filesystem {
+
+// Enumerates the different open access modes available. These are largely used
+// to parameterize types in order to constrain which API subset is available.
+enum class OpenAccess {
+  ReadOnly = O_RDONLY,
+  WriteOnly = O_WRONLY,
+  ReadWrite = O_RDWR,
+};
+
+// The different creation styles available when opening a file or directory.
+//
+// Because these are by far the most common parameters and they have unambiguous
+// names, the enumerators are also available directly within the namespace.
+enum class CreationFlags {
+  OpenExisting = 0,
+  OpenAlways = O_CREAT,
+  CreateAlways = O_CREAT | O_TRUNC,
+  CreateNew = O_CREAT | O_EXCL,
+};
+using CreationFlags::CreateAlways;
+using CreationFlags::CreateNew;
+using CreationFlags::OpenAlways;
+using CreationFlags::OpenExisting;
+
+// General flags to control the behavior of opening files that aren't covered by
+// other more specific flags.
+enum class OpenFlags {
+  None = 0,
+  Append = O_APPEND,
+  Truncate = O_TRUNC,
+};
+
+// Flags controlling which permissions should be checked in an `Access` call.
+//
+// These permissions can also be combined with the `|` operator, so `R_OK |
+// W_OK` checks for both read and write access.
+enum class AccessCheck : int {
```

Comment:
Ironically, this one is bitflags but isn't called as such, even though the comment says it. Maybe `AccessCheckFlags`?
```suggestion
enum class AccessCheckFlags : int {
```

Reply:
Done.